File: go-sdk/cortex-mem-go/dto/experience.go
package dto

// ExperienceRequest retrieves relevant experiences from memory.
// POST /api/memory/experiences
//
// Wire format (verified against backend MemoryController.java):
//   {"task":"...", "project":"/path", "count":4, "source":"...", "requiredConcepts":[...], "userId":"..."}
type ExperienceRequest struct {
	Task             string   `json:"task"`
	Project          string   `json:"project,omitempty"`        // Wire: "project" (not "project_path"); omitempty matches Java SDK behavior
	Count            int      `json:"count,omitempty"`
	Source           string   `json:"source,omitempty"`
	RequiredConcepts []string `json:"requiredConcepts,omitempty"` // Wire: camelCase!
	UserID           string   `json:"userId,omitempty"`           // Wire: camelCase!
}

// Experience represents a retrieved experience from the backend.
type Experience struct {
	ID              string  `json:"id"`
	Task            string  `json:"task"`
	Strategy        string  `json:"strategy"`
	Outcome         string  `json:"outcome"`
	ReuseCondition  string  `json:"reuseCondition"`
	QualityScore    float32 `json:"qualityScore"`
	CreatedAt       string  `json:"createdAt,omitempty"`
}

// ICLPromptRequest builds an ICL prompt from historical experiences.
// POST /api/memory/icl-prompt
//
// Wire format (verified against backend MemoryController.java):
//   {"task":"...", "project":"/path", "maxChars":4000, "userId":"..."}
type ICLPromptRequest struct {
	Task     string `json:"task"`
	Project  string `json:"project,omitempty"`     // Wire: "project" (not "project_path"); omitempty matches Java SDK behavior
	MaxChars int    `json:"maxChars,omitempty"`    // Wire: camelCase!
	UserID   string `json:"userId,omitempty"`      // Wire: camelCase!
}

// ICLPromptResult is the result from the ICL prompt builder.
type ICLPromptResult struct {
	Prompt          string `json:"prompt"`
	ExperienceCount string `json:"experienceCount"`
}

File: js-sdk/cortex-mem-js/src/dto/experience.ts
// ============================================================
// Experience & ICL DTOs
// ============================================================

import { safeStringOr, safeNumber, safeNumberOr, safeString, firstNonNullOr } from './wire-helpers';

/**
 * Request to retrieve relevant experiences.
 * POST /api/memory/experiences
 *
 * Wire format: requiredConcepts and userId are camelCase.
 */
export interface ExperienceRequest {
  task: string;
  project?: string;
  count?: number;
  source?: string;
  requiredConcepts?: string[];
  userId?: string;
}

/**
 * A retrieved experience from the backend.
 *
 * Wire format uses SNAKE_CASE (backend Jackson naming strategy).
 * Field names are normalized to camelCase via parseExperience.
 */
export interface Experience {
  id: string;
  task: string;
  strategy: string;
  outcome: string;
  /** Parsed from wire field "reuse_condition" */
  reuseCondition: string;
  /** Parsed from wire field "quality_score" */
  qualityScore: number;
  /** Parsed from wire field "created_at" */
  createdAt?: string;
}

/**
 * Parse a raw wire-format experience into the canonical Experience type.
 * Uses safe type conversion to handle null values and type mismatches gracefully.
 */
export function parseExperience(raw: Record<string, unknown>): Experience {
  return {
    id: safeStringOr(raw.id, ''),
    task: safeStringOr(raw.task, ''),
    strategy: safeStringOr(raw.strategy, ''),
    outcome: safeStringOr(raw.outcome, ''),
    reuseCondition: safeStringOr(firstNonNullOr(raw, ['reuse_condition', 'reuseCondition']), ''),
    qualityScore: safeNumberOr(firstNonNullOr(raw, ['quality_score', 'qualityScore']), 0),
    createdAt: safeString(firstNonNullOr(raw, ['created_at', 'createdAt'])),
  };
}

/**
 * Request to build an ICL prompt.
 * POST /api/memory/icl-prompt
 *
 * Wire format: maxChars and userId are camelCase.
 */
export interface ICLPromptRequest {
  task: string;
  project?: string;
  maxChars?: number;
  userId?: string;
}

/**
 * Result from the ICL prompt builder.
 *
 * Use {@link parseICLPromptResult} to safely parse from wire format.
 */
export interface ICLPromptResult {
  prompt: string;
  experienceCount: number;
  maxChars?: number;
}

/**
 * Parse a raw wire-format ICL prompt result into the canonical type.
 * Handles null values and type mismatches gracefully.
 */
export function parseICLPromptResult(raw: Record<string, unknown>): ICLPromptResult {
  return {
    prompt: safeStringOr(raw.prompt, ''),
    experienceCount: safeNumberOr(firstNonNullOr(raw, ['experienceCount', 'experience_count']), 0),
    maxChars: safeNumber(firstNonNullOr(raw, ['maxChars', 'max_chars'])),
  };
}
